calibrators/intrinsic_camera_calibrator/ceres_intrinsic_camera_calibrator/include/ceres_intrinsic_camera_calibrator/coefficients_residual.hpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 Tier IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CERES_INTRINSIC_CAMERA_CALIBRATOR__COEFFICIENTS_RESIDUAL_HPP_
+#define CERES_INTRINSIC_CAMERA_CALIBRATOR__COEFFICIENTS_RESIDUAL_HPP_
+
+#include <Eigen/Core>
+#include <opencv2/core.hpp>
+
+#include <ceres/autodiff_cost_function.h>
+#include <ceres/ceres.h>
+
+struct CoefficientsResidual
+{
+  static constexpr int RESIDUAL_DIM = 8;
+
+  CoefficientsResidual(
+    int radial_distortion_coeffs, bool use_tangential_distortion, int rational_distortion_coeffs,
+    int num_samples_factor, double regularization_weight)
+  {
+    radial_distortion_coeffs_ = radial_distortion_coeffs;
+    use_tangential_distortion_ = use_tangential_distortion;
+    rational_distortion_coeffs_ = rational_distortion_coeffs;
+    num_samples_factor_ = num_samples_factor;
+    regularization_weight_ = regularization_weight;
+  }
+
+  /*!
+   * The cost function representing the reprojection error
+   * @param[in] camera_intrinsics The camera intrinsics
+   * @param[in] residuals The residual error of projecting the tag into the camera
+   * @returns success status
+   */
+  template <typename T>
+  bool operator()(const T * const camera_intrinsics, T * residuals) const
+  {
+    const T null_value = T(0.0);
+    int distortion_index = 4;
+    const T & k1 =
+      radial_distortion_coeffs_ > 0 ? camera_intrinsics[distortion_index++] : null_value;
+    const T & k2 =
+      radial_distortion_coeffs_ > 1 ? camera_intrinsics[distortion_index++] : null_value;
+    const T & k3 =
+      radial_distortion_coeffs_ > 2 ? camera_intrinsics[distortion_index++] : null_value;
+    const T & p1 = use_tangential_distortion_ ? camera_intrinsics[distortion_index++] : null_value;
+    const T & p2 = use_tangential_distortion_ ? camera_intrinsics[distortion_index++] : null_value;
+    const T & k4 =
+      rational_distortion_coeffs_ > 0 ? camera_intrinsics[distortion_index++] : null_value;
+    const T & k5 =
+      rational_distortion_coeffs_ > 1 ? camera_intrinsics[distortion_index++] : null_value;
+    const T & k6 =
+      rational_distortion_coeffs_ > 2 ? camera_intrinsics[distortion_index++] : null_value;
+
+    residuals[0] = num_samples_factor_ * regularization_weight_ * pow(k1, 2);
```

Comment:
@amadeuszsz 
Please have a look at:
http://ceres-solver.org/nnls_tutorial.html#hello-world

I am very sure the pow is not needed here

Reply:
Also, the `num_samples` and `regularization_weight` are being squared.
As far as I know, the way to apply weights is something like:
https://github.com/tier4/CalibrationTools/blob/dc1bd3a677e10a89e4324135083d0187c18b2abb/calibrators/tag_based_sfm_calibrator/src/ceres/calibration_problem.cpp#L690

Reply:
Addressed in https://github.com/tier4/CalibrationTools/pull/208/commits/2e26b158b248757abd99ce41ca6f9a013c77a6fa